# Rebrickable - Lego Data Analysis

Sourse: https://rebrickable.com/downloads/

In [2]:
import pandas as pd
import numpy as np

In [3]:
# Setting the Data Frames
sets = pd.read_csv('Datasets/sets.csv')
themes = pd.read_csv('Datasets/themes.csv')
inventory_parts_df = pd.read_csv('Datasets/inventory_parts.csv')
color_df = pd.read_csv('Datasets/colors.csv')

### Display Top 10 Rows of the sets dataset

In [4]:
sets.head(10)

set_num                            name  year  theme_id  num_parts
0   001-1                           Gears  1965         1         43
1  0011-2               Town Mini-Figures  1979        67         12
2  0011-3      Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1              Space Mini-Figures  1979       143         12
4  0013-1              Space Mini-Figures  1979       143         12
5  0014-1              Space Mini-Figures  1979       143         12
6  0015-1              Space Mini-Figures  1979       143         18
7  0016-1             Castle Mini Figures  1979       186         15
8   002-1  4.5V Samsonite Gears Motor Set  1965         1          3
9   003-1             Master Mechanic Set  1966       366        403

### Display Last 10 Rows of the sets datasets

In [5]:
sets.tail(10)

set_num                              name  year  theme_id  \
18956   XBOXGAME-1   Batman the Videogame - Xbox 360  2008       501   
18957   XBOXGAME-2        Bionicle Heroes - Xbox 360  2006       501   
18958   XBOXGAME-3      DC Super-Villains - Xbox One  2018       501   
18959   XBOXGAME-9  Star Wars: The Video Game - Xbox  2005       501   
18960   XMASTREE-1                    Christmas Tree  2019       410   
18961      XWING-1               Mini X-Wing Fighter  2019       158   
18962      XWING-2                 X-Wing Trench Run  2019       158   
18963  YODACHRON-1   Yoda Chronicles Promotional Set  2013       158   
18964   YTERRIER-1                 Yorkshire Terrier  2018       598   
18965     ZX8000-1              ZX 8000 LEGO Sneaker  2020       501   

       num_parts  
18956          0  
18957          0  
18958          0  
18959          0  
18960         26  
18961         60  
18962         52  
18963        413  
18964          0  
18965          0

### Check Datatype of Each Column

In [6]:
sets.dtypes

set_num      object
name         object
year          int64
theme_id      int64
num_parts     int64
dtype: object

### Check Null Values in Sets and Themes

In [7]:
sets.isnull().sum()

set_num      0
name         0
year         0
theme_id     0
num_parts    0
dtype: int64

In [8]:
themes.isnull().sum()

id             0
name           0
parent_id    138
dtype: int64

### How many Rows and Columns are there in datasets

In [9]:
# (Rows, Columns)
sets.shape

(18966, 5)

In [10]:
sets.columns

Index(['set_num', 'name', 'year', 'theme_id', 'num_parts'], dtype='object')

In [11]:
len(sets.columns)

5

In [12]:
len(sets)

18966

In [13]:
sets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18966 entries, 0 to 18965
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   set_num    18966 non-null  object
 1   name       18966 non-null  object
 2   year       18966 non-null  int64 
 3   theme_id   18966 non-null  int64 
 4   num_parts  18966 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 741.0+ KB


In [14]:
themes.shape

(444, 3)

### Merge Data Sets and Themes

In [15]:
sets_themes_df = pd.merge(sets,themes, how='left', left_on='theme_id', right_on='id',suffixes=('_sets','_themes'))
sets_themes_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18966 entries, 0 to 18965
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   set_num      18966 non-null  object 
 1   name_sets    18966 non-null  object 
 2   year         18966 non-null  int64  
 3   theme_id     18966 non-null  int64  
 4   num_parts    18966 non-null  int64  
 5   id           18966 non-null  int64  
 6   name_themes  18966 non-null  object 
 7   parent_id    8676 non-null   float64
dtypes: float64(1), int64(4), object(3)
memory usage: 1.3+ MB


In [16]:
sets_themes_df.head(9)

set_num                       name_sets  year  theme_id  num_parts   id  \
0   001-1                           Gears  1965         1         43    1   
1  0011-2               Town Mini-Figures  1979        67         12   67   
2  0011-3      Castle 2 for 1 Bonus Offer  1987       199          0  199   
3  0012-1              Space Mini-Figures  1979       143         12  143   
4  0013-1              Space Mini-Figures  1979       143         12  143   
5  0014-1              Space Mini-Figures  1979       143         12  143   
6  0015-1              Space Mini-Figures  1979       143         18  143   
7  0016-1             Castle Mini Figures  1979       186         15  186   
8   002-1  4.5V Samsonite Gears Motor Set  1965         1          3    1   

    name_themes  parent_id  
0       Technic        NaN  
1  Classic Town       50.0  
2  Lion Knights      186.0  
3  Supplemental      126.0  
4  Supplemental      126.0  
5  Supplemental      126.0  
6  Supplemental      126.0  
7        Castle        NaN  
8       Technic        NaN

In [17]:
sets_themes_df.columns

Index(['set_num', 'name_sets', 'year', 'theme_id', 'num_parts', 'id',
       'name_themes', 'parent_id'],
      dtype='object')

### What is the total number of parts per theme?

In [18]:
num_parts_per_theme = sets_themes_df.groupby('name_themes').sum().reset_index()
num_parts_per_theme = num_parts_per_theme[['name_themes','num_parts']].sort_values('num_parts', ascending=False)
num_parts_per_theme

name_themes  num_parts
303         Technic     221214
292       Star Wars     220924
66   Creator Expert     128830
64          Creator     102357
209         Ninjago      96629
..              ...        ...
16         Aquazone         26
354         Zooters         25
229      Powered Up         21
339     Value Packs          1
6       Adventurers          0

[356 rows x 2 columns]

Answer: There are parts per theme on the list.

### What is the total number of parts per year?

In [19]:
num_parts_per_year = sets_themes_df.groupby('year').sum().reset_index()
num_parts_per_year = num_parts_per_year[['year','num_parts']].sort_values('num_parts',ascending=False)
num_parts_per_year

year  num_parts
70  2021     264837
69  2020     211871
68  2019     191559
66  2017     178609
67  2018     171723
..   ...        ...
5   1956        246
3   1954        117
8   1959         65
2   1953         50
1   1950          6

[72 rows x 2 columns]

### How many sets where created in each Century in the dataset?

We have to add Century as a derived column to the sets_themes datasets

In [20]:
unique_year = sets_themes_df['year'].unique()
unique_year.sort()
unique_year

array([1949, 1950, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961,
       1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972,
       1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983,
       1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994,
       1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,
       2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016,
       2017, 2018, 2019, 2020, 2021, 2022], dtype=int64)

In [21]:
sets_themes_df['year'] > 2000
sets_themes_df['Century'] = np.where(sets_themes_df['year'] > 2000, '21st_Century','20th_Century')
sets_themes_df

set_num                        name_sets  year  theme_id  \
0            001-1                            Gears  1965         1   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18961      XWING-1              Mini X-Wing Fighter  2019       158   
18962      XWING-2                X-Wing Trench Run  2019       158   
18963  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18964   YTERRIER-1                Yorkshire Terrier  2018       598   
18965     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501   

       num_parts   id   name_themes  parent_id       Century  
0             43    1       Technic        NaN  20th_Century  
1             12   67  Classic Town       50.0  20th_Century  
2              0  199  Lion Knights      186.0  20th_Century  
3             12  143  Supplemental      126.0  20th_Century  
4             12  143  Supplemental      126.0  20th_Century  
...          ...  ...           ...        ...           ...  
18961         60  158     Star Wars        NaN  21st_Century  
18962         52  158     Star Wars        NaN  21st_Century  
18963        413  158     Star Wars        NaN  21st_Century  
18964          0  598   Promotional        NaN  21st_Century  
18965          0  501          Gear        NaN  21st_Century  

[18966 rows x 9 columns]

In [22]:
num_sets_per_century = sets_themes_df.groupby('Century').count().reset_index()
num_sets_per_century = num_sets_per_century[['Century','set_num']]
num_sets_per_century

Century  set_num
0  20th_Century     4858
1  21st_Century    14108

In the 20th Century has 4858 sets.
And the 21st Century has 14,108 sets.

### What percentage of sets ever released in the 21st Century were Trains Themed?

In [23]:
century_21_data = sets_themes_df[sets_themes_df['Century'] == '21st_Century']
train = century_21_data[century_21_data['name_themes'].str.contains('Trains')]

In [24]:
n_century = century_21_data.shape[0]

In [25]:
n_train = train.shape[0]

Percent

In [26]:
n_train / n_century * 100

0.4182024383328608

The porcentage is 0.4182 from the Trains theme. 

### What percentage of sets ever released in the 21st Century were Disney Themed ?

In [27]:
disney_theme = century_21_data[century_21_data['name_themes'].str.contains('Disney') ]
disney_theme

set_num                   name_sets  year  theme_id  \
417                10509-1              Dusty and Chug  2013       627   
419                10510-1       Ripslinger's Air Race  2013       627   
420                10511-1     Skipper's Flight School  2013       627   
424                10515-1     Ariel's Undersea Castle  2013       640   
425                10516-1   Ariel's Magical Boat Ride  2012       640   
...                    ...                         ...   ...       ...   
17919              BELLE-1              Belle's Castle  2016       579   
17969  CINDERELLASCASTLE-1         Cinderella’s Castle  2017       608   
18772               ROSE-1  The Beast's Enchanted Rose  2017       608   
18870       TRUCOGSWORTH-1                   Cogsworth  2016       608   
18880         TRULUMIERE-1                     Lumiere  2016       608   

       num_parts   id      name_themes  parent_id       Century  
417           16  627    Disney Planes      504.0  21st_Century  
419           40  627    Disney Planes      504.0  21st_Century  
420           53  627    Disney Planes      504.0  21st_Century  
424           39  640  Disney Princess      504.0  21st_Century  
425           30  640  Disney Princess      504.0  21st_Century  
...          ...  ...              ...        ...           ...  
17919         72  579  Disney Princess        NaN  21st_Century  
17969        102  608           Disney        NaN  21st_Century  
18772          0  608           Disney        NaN  21st_Century  
18870         48  608           Disney        NaN  21st_Century  
18880         24  608           Disney        NaN  21st_Century  

[177 rows x 9 columns]

In [28]:
n_disney = disney_theme.shape[0]
n_disney

177

Percent:

In [29]:
n_disney / n_century * 100

1.2546073149985824

The porcentage is 1.25 from the Disney Theme

### What is the popular theme by year in terms of sets released in the 21st Century?

In [30]:
popular_theme_per_year = century_21_data.groupby(['year','name_themes']).count().reset_index()
popular_theme_per_year = popular_theme_per_year[['year','name_themes','set_num']].sort_values('set_num',ascending=False)
popular_theme_per_year

year   name_themes  set_num
1327  2021          Gear      226
1181  2019          Gear      173
1252  2020          Gear      150
1035  2017          Gear      141
1104  2018          Gear      115
...    ...           ...      ...
1137  2018     Spiderman        1
668   2012  Thanksgiving        1
675   2012     Valentine        1
1131  2018    Powered Up        1
538   2010  Thanksgiving        1

[1409 rows x 3 columns]

In [31]:
# Drop Duplicates
popular_theme_per_year.drop_duplicates('year').sort_values('year',ascending=False)

year name_themes  set_num
1384  2022        Gear       31
1327  2021        Gear      226
1252  2020        Gear      150
1181  2019        Gear      173
1104  2018        Gear      115
1035  2017        Gear      141
953   2016        Gear       98
866   2015        Gear       87
786   2014        Gear       97
707   2013        Gear       94
632   2012        Gear      105
566   2011        Gear      102
509   2010        Gear       59
454   2009        Gear      104
400   2008        Gear       84
343   2007        Gear      109
289   2006        Gear       75
214   2005    Bionicle       36
165   2004    Bionicle       48
109   2003    Bionicle       63
96    2002     Technic       65
6     2001    Bionicle       47

There are popular themes by years.

### What is the most produced color of lego ever in terms of quantity of parts?
Merge with color and invertory data frames

In [32]:
inventory_parts_df

inventory_id        part_num  color_id  quantity is_spare
0                   1        48379c01        72         1        f
1                   1           48395         7         1        f
2                   1  stickerupn0077      9999         1        f
3                   1         upn0342         0         1        f
4                   1         upn0350        25         1        f
...               ...             ...       ...       ...      ...
1008541        131522            4019         7         2        f
1008542        131522            4716         0         2        f
1008543        131522           6542a         8         1        f
1008544        131522            6589         7         4        f
1008545        131522            9244         7         1        f

[1008546 rows x 5 columns]

In [33]:
color_df

id                  name     rgb is_trans
0      -1             [Unknown]  0033B2        f
1       0                 Black  05131D        f
2       1                  Blue  0055BF        f
3       2                 Green  237841        f
4       3        Dark Turquoise  008F9B        f
..    ...                   ...     ...      ...
211  1085       Two-tone Silver  737271        f
212  1086            Pearl Lime  6A7944        f
213  1087            Duplo Pink  FF879C        f
214  1088          Medium Brown  755945        f
215  9999  [No Color/Any Color]  05131D        f

[216 rows x 4 columns]

In [34]:
merge_color_inventory_df = pd.merge(inventory_parts_df,color_df, how='left',left_on='color_id', right_on='id')
merge_color_inventory_df

inventory_id        part_num  color_id  quantity is_spare    id  \
0                   1        48379c01        72         1        f    72   
1                   1           48395         7         1        f     7   
2                   1  stickerupn0077      9999         1        f  9999   
3                   1         upn0342         0         1        f     0   
4                   1         upn0350        25         1        f    25   
...               ...             ...       ...       ...      ...   ...   
1008541        131522            4019         7         2        f     7   
1008542        131522            4716         0         2        f     0   
1008543        131522           6542a         8         1        f     8   
1008544        131522            6589         7         4        f     7   
1008545        131522            9244         7         1        f     7   

                         name     rgb is_trans  
0            Dark Bluish Gray  6C6E68        f  
1                  Light Gray  9BA19D        f  
2        [No Color/Any Color]  05131D        f  
3                       Black  05131D        f  
4                      Orange  FE8A18        f  
...                       ...     ...      ...  
1008541            Light Gray  9BA19D        f  
1008542                 Black  05131D        f  
1008543             Dark Gray  6D6E5C        f  
1008544            Light Gray  9BA19D        f  
1008545            Light Gray  9BA19D        f  

[1008546 rows x 9 columns]

In [35]:
merge_color_inventory_df.rename(columns={'name':'color_name'}, inplace=True)
merge_color_inventory_df

inventory_id        part_num  color_id  quantity is_spare    id  \
0                   1        48379c01        72         1        f    72   
1                   1           48395         7         1        f     7   
2                   1  stickerupn0077      9999         1        f  9999   
3                   1         upn0342         0         1        f     0   
4                   1         upn0350        25         1        f    25   
...               ...             ...       ...       ...      ...   ...   
1008541        131522            4019         7         2        f     7   
1008542        131522            4716         0         2        f     0   
1008543        131522           6542a         8         1        f     8   
1008544        131522            6589         7         4        f     7   
1008545        131522            9244         7         1        f     7   

                   color_name     rgb is_trans  
0            Dark Bluish Gray  6C6E68        f  
1                  Light Gray  9BA19D        f  
2        [No Color/Any Color]  05131D        f  
3                       Black  05131D        f  
4                      Orange  FE8A18        f  
...                       ...     ...      ...  
1008541            Light Gray  9BA19D        f  
1008542                 Black  05131D        f  
1008543             Dark Gray  6D6E5C        f  
1008544            Light Gray  9BA19D        f  
1008545            Light Gray  9BA19D        f  

[1008546 rows x 9 columns]

In [36]:
popular_color = merge_color_inventory_df.groupby('color_name').sum().reset_index()
popular_color = popular_color[['color_name','quantity']].sort_values('quantity',ascending=False)
popular_color.head(5)

color_name  quantity
1                Black    637978
59   Light Bluish Gray    379358
198              White    375882
24    Dark Bluish Gray    267069
137                Red    253801

Answer: The most color produced is Black.